doc/source/contribute.rst

```diff
@@ -79,3 +79,154 @@ Release checklist
+
+p4a will map any dependency to an internal recipe if present, and
+otherwise use pip to obtain it regularly from whatever external source.
+
+
+Install process regarding packages
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The install/build process of a p4a project, as triggered by the
+`p4a apk` command, roughly works as follows in regards to python
+packages:
+
+1. The user has specified a project folder to install. This is either
+   just a folder with python scripts and a `main.py`, or it may
+   also have a `pyproject.toml` for a more standardized install
+
+2. Dependencies are collected: they can be either specified via
+   ``--requirements`` as a list of names or pip-style URLs, or p4a
+   can optionally scan them from a project folder via the
+   pep517 library (if there is a `pyproject.toml` or `setup.py`)
+
+3. The collected dependencies are mapped to p4a's recipes if any are
+   available for them,
+   otherwise they're kept around as external regular package references
+
+4. All the dependencies mapped to recipes are built via p4a's internal
+   mechanisms to build these recipes. (This
+   may or may not indirectly use pip, depending on whether the recipe
+   wraps a python package or not and uses pip to install ro not)
+
+5. **If the user has specified to install the project in standardized
+   ways,** then the `setup.py`/whatever build system
+   of the project will be run. This happens with cross compilation set up
+   (`CC`/`CFLAGS`/... set to use the
+   proper toolchain) and a custom site-packages location.
+   The actual comand is a simple `pip install .` in the project folder
+   with some extra options: e.g. all dependencies that were already
+   installed by recipes will be pinned with a `-c` constraints file
+   to make sure pip won't install them, and build isolation will be
+   disabled via ``--no-build-isolation`` so pip doesn't reinstall
+   recipe-packages on its own.
+
+   **If the user has not specified to use standardized build approaches**,
+   p4a will simply install all the remaining dependencies that weren't
+   mapped to recipes directly and just plain copy in the user project
+   without installing. Any `setup.py` or `pyproject.toml` of the user
+   project will then be ignored in this step.
+
+6. Google's gradle is invoked to package it all up into an `.apk`
+
+
+Overall process / package relevant notes for p4a
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Here are some common things worth knowing about python-for-android's
+dealing with python packages:
+
+- Packages will work fine without a recipe if they would also build
+  on Linux ARM, don't use any API not available in the NDK if they
+  use native code, and don't use any weird compiler flags the toolchain
+  doesn't like if they use native code. The package also needs to
+  work with cross compilation
+
+- There is currently no easy way for a package to know it is being
+  cross-compiled (at least that we know of) other than examining the
```

Comment:
I'm not sure what we try to say here. But command like `file` on a shared object would give info on the target arch it's compiled to right?

Reply:
Yes but usually you'd set that via a standardized env var and have some way e.g. in setuptools to differentiate the targets or something, and we're not really enabling or aware of such a way and not recommending this to anyone. Also, most install scripts probably just don't expect to be cross compiled, and maybe they should - it was just an acknowledgement that right now most install scripts are left mostly unaware by how we set things up

Reply:
Like, this is not our own problem alone. setuptools and such probably should provide better ways too. But it seems to me that working together with setuptools/pip people this could be possibly handled better than it is right now

Reply:
> it was just an acknowledgement that right now most install scripts are left mostly unaware by how we set things up

OK I understand the purpose of the paragraph now, thanks :+1: